### PR TITLE
random: don't benchmark functions with interpolated types

### PR DESCRIPTION
**Pull request by rfourquet**
This fixes part of #124. E.g. When wanting to benchmark
`rand(Int)` and `rand(UInt)`, the following loop won't
measure what is expected:
`for T=(Int, UInt); @benchmarkable rand($T); end`
In order to benchmark the equivalent of
`@benchmarkable rand(Int); @benchmarkable rand(UInt)`,
we "value-fy" the types with `Val`.

### CODE REVIEW

**Comment by rfourquet** [comment]
Gentle bump.

**Comment by rfourquet** [comment]
cc. @jrevels @ararslan

File: src/random/RandomBenchmarks.jl [hunk 1]
```diff
@@ -56,6 +56,13 @@ set_tolerance!(g)
 
 g = addgroup!(SUITE, "types", ["rand", "rand!", "randn", "randn!", "randexp", "randexp!"])
 
+# these functions are used to interpolate values instead of Types in @benchmarkable,
+# to measure the best possible performance, cf. #124
+@inline _rand{T}(::Val{T}) = rand(Base.GLOBAL_RNG, T)
+@inline _rand{T}(rng, ::Val{T}) = rand(rng, T)
+@inline _randn{T}(rng, ::Val{T}) = randn(rng, T)
+@inline _randexp{T}(rng, ::Val{T}) = randexp(rng, T)
```
Comment (jrevels) [inline src/random/RandomBenchmarks.jl]: Could we switch this to use `where` syntax?

Other than that, this all LGTM :)

Comment (rfourquet) [reply src/random/RandomBenchmarks.jl]: Thanks!! If I remember correctly, I didn't use `where` to satisfy CI (for julia 0.5 I think). I will assume that it's not necessary anymore and will change that tomorrow (it's too late here!).

**Comment by cstjean** [comment]
This LGTM, but I'll suggest that instead of introducing the extra `_rand` functions, you could just use `@eval` on each benchmark, which will splice in the type. To take the example from #124, this:

```julia
T = UInt; @eval run(tune!(@benchmarkable rand($T)))
```

has the same runtime as `run(tune!(@benchmarkable rand(UInt))` on my system.

**Comment by cstjean** [comment]
That is - if I've understood everything correctly 🤔

**Comment by rfourquet** [comment]
I think i tried unsucessfully every combination of `@eval`, `@benchmarkable` and `@quote`... I guess the tricky part is that in the file you don't invoke directly `run (tune!(...))`. But i may have missed something obvious i will recheck.

**Comment by cstjean** [comment]
`g["rand",     "ImplicitRNG",  tstr] = @eval @benchmarkable rand($T)`?

**Comment by rfourquet** [comment]
So you seem to be right! maybe I had tried to put the eval out the loops, or maybe I dismissed this solution because it evals also the rng argument in `@eval @benchmarkable rand($MT, $T)`, which means I think that it (`MT`) is not seen by `@benchmarkable` as a "dollar-value" (the dollar is consumed by `@eval`), but apparently it doensn't make a difference in this case (`MT` is a `const`). Shall I switch to the `@eval` version @jrevels ?

**Comment by jrevels** [comment]
> Shall I switch to the @eval version @jrevels ?

Up to you. The `@eval` trick is nifty, but I think the `_rand` approach is far clearer to read.  It can get pretty hairy to understand what code is doing when splicing is used across several layers of nested macros/evaluation scopes 😛

**Comment by rfourquet** [comment]
> Up to you

Then I will keep it as is! I indeed am uncomfortable with nesting two macros for which `$` has a special meaning. Thanks!

**Pull request merged**